Reservation Rate excel export

HotStew opened the pull request:
### **Changes are well documented in commit messages and code!**

Not to be merged quite yet!

### To be done:
- [x] Tests
  * Tests are done, but tests unrelated to these changes are failing.
- [x] Make up our mind on method of time filtering

See frontend PR: https://github.com/City-of-Helsinki/huvaja/pull/338
___________
### Example of generated excel file.
![image](https://user-images.githubusercontent.com/31072611/156333279-089fd8c1-4c31-4dd4-89b3-d5cd0a192358.png)

Commit:
Create Reservation Rate Excel report

Add new endpoint where you can download an excel
report that contains a separate sheet for each
unit. Each sheet will have a summary that contains
unit name, time filters, reservation rate. Below
summary a list of resources with their sums
of reserved time per each resource. Below this
will be listings of reservation details per resource.

The endpoint will take parameters that will be used
for filtering the queries. Required filters are:
list of units, begin date and end date. Times are
optional, but will default to 08:00 and 16:00.

Create serializers for the purpose of having a simple
top-down data structure.

Use the BaseReport base class for the new view.

Create a custom excel renderer.

tuomas777 commented on reports/api/reservation_rate.py:
```diff
@@ -0,0 +1,310 @@
+        Compiles data to be presentation ready for renderer.
+        """
+
+        begin = data["begin"]
+        end = data["end"]
+        del data["begin"]
+        del data["end"]
```
```suggestion
        begin = data.pop("begin")
        end = data.pop("end")
```

HotStew replied:
Haha :D How could one forget such a simple dictionary method!

tuomas777 commented on reports/api/reservation_rate.py:
```diff
@@ -0,0 +1,310 @@
+                resource["reserved_time_sum"] = self._hour_min_string(resource_reserved_time_sum_decimal)
+
+            # Calculate max reservable time for the whole unit.
+            day_diff = end - begin; day_diff = day_diff.days
```
```suggestion
            day_diff = (end - begin).days
```

tuomas777 commented on reports/api/reservation_rate.py:
```diff
@@ -0,0 +1,310 @@
+                
+                row_pos += 4
+
+            row_pos = 0
```
I think it would make this thing easier to figure out if this was set at the start of each iteration.

tuomas777 commented on reports/api/reservation_rate.py:
```diff
@@ -0,0 +1,310 @@
+        return context
+
+    def get_filename(self, request, validated_data):
+        return 'varaus_aste_raportti.xlsx'
```
Minor of course, but as the word is a closed compound `varausaste`, I think it should be `varausaste` in the filename as well. Hurts my eyes otherwise 😄

tuomas777 commented on reports/api/reservation_rate.py:
```diff
@@ -0,0 +1,310 @@
+    
+                for reservation in resource["reservations"]:
+                    resource_reserved_time_sum_decimal += reservation["reserved_time"]
+                    del reservation["reserved_time"]
```
```suggestion
                resource_reserved_time_sum_decimal = sum(r.pop("reserved_time") for r in resource["reservations"])
```

tuomas777 commented on reports/api/reservation_rate.py:
```diff
@@ -0,0 +1,310 @@
+            end = datetime.datetime.strptime(
+                f"{end_date} {end_time}", "%Y-%m-%d %H:%M"
+            )
+        except Exception as e:
```
Would be better if the exact exception(s) were caught here instead.

tuomas777 commented on reports/tests/test_reservation_rate.py:
```diff
@@ -0,0 +1,56 @@
+    response = api_client.get(
+        f"{url}?units={reservation.resource.unit.id}&start_date=3-04-01&end_date=2015-04-06&start_time=08:00&end_time=16:00"
+    )
+    assert response.status_code == 400
```
I would also check error messages in addition to status codes on those assertions. There are a ton of reasons why a request might return 404 or 400, so these could be passing falsely. Also, a horrible mistake on this line: a missing newline!

Commit:
Remove impossible test

Test kept failing due to it testing that an exception
should be raised. But the code that it tests is now
wrapped in a try except block which swallows the raised
exception so it can not be passed to the test.

HotStew commented:
Most of the tests that were failing were due to a translation mismatch. So nothing critical and doesn't have to be bothered with in this PR. I did however remove another test that was failing due to code changes from a ~year ago that made the test void.

frwickst reviewed (approved):
Quite a lot of things going on here, but I will approve this. I don't know the business logic here or if the data itself is the correct one, I mainly looked at the code in general.

HotStew commented:
Added translations to API error messages for fields that the user selects on the frontend.
